src/main/scala/zio/keeper/Cluster.scala

```diff
@@ -1,31 +1,355 @@
 package zio.keeper
 
-import java.net.InetAddress
+import java.math.BigInteger
+import java.util.UUID
 
-import zio.{ IO, ZIO }
+import zio._
+import zio.console.putStrLn
+import zio.keeper.Cluster.Transport.TCPTransport
+import zio.keeper.Cluster.{ Credentials, Discovery }
+import zio.keeper.EchoServerProtocol.EchoMessage
+import zio.keeper.Error.{ HandshakeError, SendError }
+import zio.keeper.Error.ProtocolDef.NodeUnknown
+import zio.nio._
+import zio.nio.channels.{ AsynchronousServerSocketChannel, AsynchronousSocketChannel }
 import zio.stream.Stream
-import zio.keeper.Cluster._
 
-trait Cluster {
-  final def join: ZIO[Credentials with Discovery with Transport, Error, Unit] = ???
+import scala.reflect.ClassTag
+
+trait Cluster[A] {
+  def nodes: UIO[List[NodeId]]
+
+  def send[B <: A](data: B, receipt: NodeId): IO[Error, Unit]
+
+  def broadcast[B <: A](data: B)(implicit classTag: ClassTag[B]): IO[Error, Unit]
+
+  def receive: Stream[Error, Message[A]]
 }
 
 object Cluster {
 
+  sealed trait InternalProtocol
```

Comment:
I'd separate it from the notion of protocol, to simplify naming. For example:

```scala
sealed trait Payload
object Payload {
  // TODO: define them
}

// TODO: see the remark below regarding serdes
val protocol = new Protocol[Event] { ... }
```
Visibility can be tuned, but I'd pick `private` wherever possible. The more I think about it, the more it looks as `Protocol` should be renamed to `Schema`.

Reply:
done